### PR TITLE
Topic/jasonlu/dict clone performance

### PR DESCRIPTION
**Pull request by jasonlue**
Fix Dict clone performance issue when Dict is 1M+. Please refer https://github.com/zeek/zeek/issues/2387 for more details

This is the core pull request extracted from https://github.com/zeek/zeek/pull/2456. It passed all btests. PR 2456 will be closed without merge. 
 
Detects performance deterioration by monitoring average distance to find the next space to insert the new item. When the space is hard to find, it means the load factor is just too high at the moment. SizeUp the dictionary (double the size) when this happens.
Fine tunes the threshold for performance deterioration: The condition is:
(1) The current load factor is high ( > 25%)
(2) Average distance to find a space for insertion is long (> 32)
(3) Enough samples are collected: (> 128 times.)
@timwoj @JustinAzoff @awelzel

### CODE REVIEW

File: src/Dict.h [hunk 1]
```diff
@@ -1648,6 +1672,9 @@ template <typename T> class Dictionary
 	uint32_t num_entries = 0;
 	uint32_t max_entries = 0;
 	uint64_t cum_entries = 0;
+	int64_t space_distance_samples = 0;
```
Comment (awelzel) [inline src/Dict.h:1675]: They can probably be unsigned. I'm tempted to say that 32bit is probably enough and overflows are not a correctness issue, too. But maybe the 8bytes aren't worth the discussion ;-)

File: src/Dict.h [hunk 2]
```diff
@@ -639,6 +650,17 @@ template <typename T> class Dictionary
...
 				SizeUp();
+			// if space_distance is too great, performance decreases. we need to sizeup for
+			// performance.
+			else if ( space_distance_samples > detail::MIN_SPACE_DISTANCE_SAMPLES && 
+				space_distance_sum > space_distance_samples * detail::SPACE_DISTANCE_THRESHOLD &&
+			          int(num_entries) > detail::MIN_DICT_LOAD_FACTOR_100 * Capacity() / 100 )
+				{
+				SizeUp();
+				// reset performance metrics.
+				space_distance_sum = 0;
```
Comment (awelzel) [inline src/Dict.h:661]: Should these distance metrics be invalidated after the other `SizeUp()` calls as well?

Mostly wondering about corner cases where SizeUp() happens due to ThresholdEntries() (e.g. line 651) and then another one due to the distance metrics which aren't valid for the new size of the dictionary anymore. Also generally seems more consistent to reset these metrics unconditionally after resizing (maybe even reset in SizeUp() ? ).

Comment (jasonlue) [reply src/Dict.h:661]: Good idea. To avoid the consecutive, probably unnecessary SizeUp().

**Comment by timwoj** [comment]
This definitely looks a lot better in regards to @awelzel's original test case from #2387:

```
Size, 1000000, fill, 0.580525, iterate, 0.476089, copy, 0.328293
Size, 1100000, fill, 0.641841, iterate, 0.528706, copy, 0.512816
Size, 1200000, fill, 0.69377, iterate, 0.587709, copy, 0.558212
Size, 1300000, fill, 0.752502, iterate, 0.637213, copy, 0.60088
Size, 1400000, fill, 0.81399, iterate, 0.688173, copy, 0.654635
Size, 1500000, fill, 0.878005, iterate, 0.750905, copy, 0.723649
Size, 1600000, fill, 0.980179, iterate, 0.836262, copy, 0.809011
Size, 1700000, fill, 1.091124, iterate, 0.888452, copy, 0.809545
Size, 1800000, fill, 1.126841, iterate, 0.946189, copy, 0.896005
Size, 1900000, fill, 1.197061, iterate, 1.055226, copy, 0.929312
Size, 2000000, fill, 1.242657, iterate, 1.10663, copy, 1.040759
```

**Comment by jasonlue** [comment]
@timwoj 

Please let me what needs to be done before the PR closes.

**Comment by timwoj** [comment]
> Please let me know what needs to be done before the PR closes.

I'm waiting to make sure that CI passes through, and I'd like to run some manual benchmarks I have on it before approving. I'm having some issues with access to our benchmarking hosts right now so it might not happen until next week.

**Comment by jasonlue** [comment]
Sure.

**Comment by timwoj** [comment]
My usual suite of benchmarks doesn't show anything really unusual here. Most of the numbers are +/- 1% of what the numbers from master are, and it fixes the performance issue in the reported issue.

@awelzel Did you have any other comments?

**Review by awelzel — APPROVED** [review]
> @awelzel Did you have any other comments?

Nope.

**Pull request merged**